redash/schema.py

```diff
@@ -0,0 +1,186 @@
+    tables_to_create = existing_tables_set.difference(persisted_table_set)
+
+    table_metadata = [table_data[table_name] for table_name in tables_to_create]
+
+    models.db.session.bulk_insert_mappings(TableMetadata, table_metadata)
+
+
+def insert_or_update_column_metadata(table, existing_columns_set, column_data):
+    persisted_columns = ColumnMetadata.query.filter(
+        ColumnMetadata.name.in_(existing_columns_set),
+        ColumnMetadata.table_id == table.id,
+    ).all()
+
+    persisted_column_data = []
+    for persisted_column in persisted_columns:
+        # Add IDs to persisted column data so it can be used for updates.
+        column_data[persisted_column.name]["id"] = persisted_column.id
+        persisted_column_data.append(column_data[persisted_column.name])
+
+    models.db.session.bulk_update_mappings(ColumnMetadata, persisted_column_data)
+
+    # Find the columns that need to be created by subtracting the sets:
+    persisted_column_set = set([col_data["name"] for col_data in persisted_column_data])
+    columns_to_create = existing_columns_set.difference(persisted_column_set)
+
+    column_metadata = [column_data[col_name] for col_name in columns_to_create]
+
+    models.db.session.bulk_insert_mappings(ColumnMetadata, column_metadata)
+
+
+class SchemaCache(object):
+    """
+    This caches schema requests in redis and uses a method to
+    serve stale values while the cache is being populated or
+    updated to handle the thundering herd problem.
+    """
+
+    timeout = settings.SCHEMAS_REFRESH_SCHEDULE * 60
+    # keeping the stale cached items for 10 minutes longer
+    # than its timeout to make sure repopulation can work
+    stale_cache_timeout = 60 * 10
+
+    def __init__(self, data_source):
+        self.data_source = data_source
+        self.client = redis_connection
+        self.cache_key = "data_source:schema:cache:{}".format(self.data_source.id)
+        self.lock_key = "{}:lock".format(self.cache_key)
+        self.fresh_key = "{}:fresh".format(self.cache_key)
+
+    def load(self):
+        """
+        When called will fetch all table and column metadata from
+        the database and serialize it with the TableMetadataSerializer.
+        """
+        # due to the unfortunate import time side effects of
+        # Redash's package layout this needs to be done inline
+        from redash.serializers import TableMetadataSerializer
+
+        schema = []
+        tables = (
+            TableMetadata.query.filter(
+                TableMetadata.data_source_id == self.data_source.id,
+                TableMetadata.exists.is_(True),
+            )
+            .order_by(TableMetadata.name)
+            .options(
+                joinedload(TableMetadata.columns),
+                joinedload(TableMetadata.sample_queries),
+            )
+        )
+
+        for table in tables:
+            schema.append(
+                TableMetadataSerializer(table, with_favorite_state=False).serialize()
```

Comment:
Is `with_favorite_state` supposed to be there? I'm not sure what this is used for?

Reply:
So behind the scenes the `TableMetadataSerializer` calls `serialize_table_metadata` by passing in the kwargs as options for the `QuerySerializer` to serialize the sample queries. That serializers uses the same options to decide whether to current user has [marked one of the sample queries as their favorite](https://github.com/mozilla/redash/blob/b839d32ab234a8f220d8c8fd21d9ae1b7b429ad1/redash/serializers/__init__.py#L75). Which I guessed is an unneeded database query for our use case here, it shaved off a few seconds.